### ISSUE TITLE
Vocab size mismatch

### ISSUE THREAD

**Issue by deeptibhegde** [rd20karim/M2T-Segmentation#2]
I processed the data as mentioned in the README and get the following error when I run 

     python visualizations/poses2concepts.py --config ./configs_kit_aug/MLP.yaml --dataset_name kit --n_map 5 --n_gifs 50 --save_results ./gifs_map_kit22_

ERROR MESSAGE:

        2024-05-01 18:12:11,523 [INFO] Building dataset ... 
        2024-05-01 18:12:11,823 [INFO] CORRECT TOKENS False
        2024-05-01 18:12:12,073 [INFO] Building vocabulary with minimum frequency : 3
        2024-05-01 18:12:12,422 [INFO] Std/Mean Normalization
        2024-05-01 18:12:13,426 [INFO] Convert token to numerical values
        2024-05-01 18:12:13,767 [INFO] Using official split
        2024-05-01 18:12:13,853 [INFO] Number  of samples VAL: 470 TRAIN: 7451, TEST: 1376 
        2024-05-01 18:12:16,133 [INFO] VOCAB SIZE  = 1709 
        2024-05-01 18:12:17,974 [INFO] Scaling 1.0
        2024-05-01 18:12:17,981 [INFO] Applying local_recurrent attention
        Traceback (most recent call last):
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/visualizations/poses2concepts.py", line 180, in <module>
            loaded_model, train_data_loader, test_data_loader,val_data_loader = load_model_config(device=device,args=args)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/src/evaluate_m2L.py", line 76, in load_model_config
            loaded_model.load_state_dict(new_dict)
          File "/home/dhegde/anaconda3/envs/llava/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2041, in load_state_dict
            raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
        RuntimeError: Error(s) in loading state_dict for seq2seq:
                size mismatch for dec.embedding.weight: copying a param with shape torch.Size([878, 64]) from checkpoint, the shape in current model is torch.Size([1709, 64]).
                size mismatch for dec.fc.weight: copying a param with shape torch.Size([878, 192]) from checkpoint, the shape in current model is torch.Size([1709, 192]).
                size mismatch for dec.fc.bias: copying a param with shape torch.Size([878]) from checkpoint, the shape in current model is torch.Size([1709]).

**Comment by rd20karim** [comment]
I think the problem is in the dataset correction step  that you mentioned in the [first issue](https://github.com/rd20karim/M2T-Segmentation/issues/1#issue-2274408585). I can't reproduce this error, can you check the number of samples in step 1 [build_data.py](https://github.com/rd20karim/M2T-Segmentation/blob/f9b60b40f16e709c360bb167e23470457d2f00c7/datasets/build_data.py#L61). It should be 6016 pairs of motion samples (kitmld_array) and descriptions (old_desc) in the case of KIT-ML-aug (augmented).

**Comment by deeptibhegde** [comment]
There were actually a couple issues with running build_data.py. I am a little confused with the data pre-processing and hoping you can clear it up to make sure I am on the right track:

1. What is exactly the difference between HumanML3D and KIT-ML? From [this repo](https://github.com/EricGuo5513/HumanML3D), it is my understanding that HumanML3D just consists of all the datasets included in AMASS, including KIT. If this is the case, why are there different formats (such as 22 joints in HumanML3D and 21 in KIT, and different joint connectivity)? Also, what is the difference between all the various versions of KIT such as ```kit_with_splits_2023``` and ```kitmld_anglejoint_2016_30s_final_cartesian```?
2. What is the way to get the Augmented datasets? I just followed the steps in the [linked repo](https://github.com/EricGuo5513/HumanML3D), but I wanted to generate just the KIT data. I still get 22 joints. Is it enough to just select the first 21 joints to use ```kit_m2t_dataset.py```? 
3. How can I preprocess any SMPL model to run a demo to perform motion-to-text? I have tried to use the HumanML3D pre-processing steps, but am confused as to which model/dataloader to use because of the vocab size mismatch.
4. Another thing I noticed was the difference in scales of the joints. The pre-processed KIT data linked in the ```README``` , the ```['normalized_poses']``` is of the scale 10^-3, whereas when I pre-process the data myself, I get the scale as 10^-1. Could you explain how I can take any sequence of SMPLs and bring it to the required format? I think there is an issue with normalization and joint order I am not seeing.

I believe the answers to these questions should allow me to know how to properly load the appropriate data and model. 

Thanks!

**Comment by rd20karim** [comment]
1. KIT-ML and HumanML3D are two different datasets. The original KIT version was released in 2016 by plappert et al. Then Guo et al. created an augmented version of this dataset along with HumanMl3D (released in 2022), the main experiments were focused on original KIT version as described in the [paper](https://link.springer.com/epdf/10.1007/s00521-023-09227-z?sharing_token=YrUK4rjCVvroLcdPW_wgffe4RwlQNchNByi7wbcMAY78SE_ehDTlEAMngsMFB353ZcAy13L8K6m_ZAM36z-5O-Fom-ddsCq6QHCUvgA6G1RNL_LspuHgHgz4FTJcVUjVilF3Zd872mHtx2NUdAMBaAyXDI0OGIckD5K32F6oYoc%3D), then I extended results to this new versions after their release.

-   Original KIT= `kitmld_anglejoint_2016_30s_final_cartesian`
-   Augmented KIT-ML= `kit_with_splits_2023` (Augmented from the Original KIT by mirroring motions and changing descriptions accordingly (e.g., left->right ..)
-   HumanML3D= Amass+HumanAct12
-   Amass= Unify multiple datasets by mapping their sparse markers (Motion Capture data) to SMPL body models that has 22 joints.

2. By following exactly  the steps in the linked repo you can get both datasets HumanML3D and KIT (augmented), if you are interested only on KIT-ML-aug you can directly download it from their link [here](https://github.com/EricGuo5513/HumanML3D?tab=readme-ov-file#how-to-obtain-the-data)

3. HumanML3D already provided pre-processed SMPLs mapped to 22 joints unifying multiple datasets. After getting this dataset from the repo you can just follow the steps mentioned [here](https://github.com/rd20karim/M2T-Segmentation?tab=readme-ov-file#preprocess-dataset)

4. The datalink I provided in the readme contains already pre-processed data for original KIT version to simplify the process and avoid errors. Data for the others needs to be obtained from the main sources. The Original KIT data doesn't require build_data.py and results could be obtained directly by setting correct path as described in the Readme.

Feel free to ask if you have any further questions!

**Comment by deeptibhegde** [comment]
Gotcha, thanks!


When I try to run ```build.py``` with the data provided by the HumanML3D dataset, I get the below error:


      Traceback (most recent call last):
        File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/build_data.py", line 73, in <module>
          np.savez(r"/home/dhegde/projects/motion-pipe/M2T-Segmentation/data/kit_with_splits_2023.npz",kitmld_array=kit_poses,old_desc=list_desc,sample_ids = nameid, splits_ids=split_ids)
        File "/home/dhegde/.local/lib/python3.10/site-packages/numpy/lib/npyio.py", line 639, in savez
          _savez(file, args, kwds, False)
        File "/home/dhegde/.local/lib/python3.10/site-packages/numpy/lib/npyio.py", line 740, in _savez
          val = np.asanyarray(val)
      ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (6016,) + inhomogeneous part.

I just unzipped the data linked[ here ](https://github.com/EricGuo5513/HumanML3D?tab=readme-ov-file#how-to-obtain-the-data) and ran ```build_data.py```. Did I miss a step?

**Comment by deeptibhegde** [comment]
Okay, I fixed that error by downgrading numpy and saving the array as ```dtype=object```, but now I am getting a similar error as before: 


        [INFO] CORRECT TOKENS True
        [INFO] INITIAL CORRECTIONS AND LOWER CASING ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Token index 30 corrected cm ----> am 
        [INFO] Replacing ...
        [INFO] Token index 31 corrected sidejump ----> None 
        Traceback (most recent call last):
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 203, in <module>
            data = dataset_class(path, filter_data=True,min_freq=3)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 46, in __init__
            self.lang = vocabulary(self.sentences, correct_tokens=correct_tokens, ask_user=False)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 94, in __init__
            self.token_correction(ask_user)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 159, in token_correction
            desc = " ".join(tokens)
        TypeError: sequence item 4: expected str instance, NoneType found



Again, I am just trying to run the pre-process data on the KIT-Aug data that is available for download in the HumanML3d Repo

**Comment by rd20karim** [comment]
I updated the file [build_data.py](https://github.com/rd20karim/M2T-Segmentation/blob/main/datasets/build_data.py) along with other additional minor changes. You can try `git pull` and let me know if it works!

**Comment by rd20karim** [comment]
You can now maintain same numpy version `1.26.1` the problem of shapes should be resolved after git pull.

**Comment by deeptibhegde** [comment]
Thanks, that fixed the shape error. 

Is line 4 in vocabulary.py necessary? I get an import error at 

```from datasets.kit_dataset import kitDataset```

This file does not exist and is not used anywhere else. I have commented that line. However, I still get the same error as before when running ```python datasets/kit_m2t_dataset.py``` :

        [INFO] CORRECT TOKENS True
        [INFO] INITIAL CORRECTIONS AND LOWER CASING ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Replacing ...
        [INFO] Token index 30 corrected cm ----> am 
        [INFO] Replacing ...
        [INFO] Token index 31 corrected sidejump ----> None 
        Traceback (most recent call last):
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 203, in <module>
            data = dataset_class(path, filter_data=True,min_freq=3)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 46, in __init__
            self.lang = vocabulary(self.sentences, correct_tokens=correct_tokens, ask_user=False)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 94, in __init__
            self.token_correction(ask_user)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 159, in token_correction
            desc = " ".join(tokens)
        TypeError: sequence item 4: expected str instance, NoneType found

**Comment by rd20karim** [comment]
Yes, you can just delete this line 4, I will remove it in the next push.

**Comment by deeptibhegde** [comment]
> Thanks, that fixed the shape error.
> 
> Is line 4 in vocabulary.py necessary? I get an import error at
> 
> `from datasets.kit_dataset import kitDataset`
> 
> This file does not exist and is not used anywhere else. I have commented that line. However, I still get the same error as before when running `python datasets/kit_m2t_dataset.py` :
> 
> ```
>     [INFO] CORRECT TOKENS True
>     [INFO] INITIAL CORRECTIONS AND LOWER CASING ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Replacing ...
>     [INFO] Token index 30 corrected cm ----> am 
>     [INFO] Replacing ...
>     [INFO] Token index 31 corrected sidejump ----> None 
>     Traceback (most recent call last):
>       File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 203, in <module>
>         data = dataset_class(path, filter_data=True,min_freq=3)
>       File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/kit_m2t_dataset.py", line 46, in __init__
>         self.lang = vocabulary(self.sentences, correct_tokens=correct_tokens, ask_user=False)
>       File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 94, in __init__
>         self.token_correction(ask_user)
>       File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/vocabulary.py", line 159, in token_correction
>         desc = " ".join(tokens)
>     TypeError: sequence item 4: expected str instance, NoneType found
> ```

Okay this was resolved by using the listed version of ```pyspellchecker==0.6.2``` .

**Comment by deeptibhegde** [comment]
Thank you for your help, I was able to run inference on the KIT-ML-Augmented data that is given for download.

In order to run inference on the HumanML3D dataset, do I need to download and process the entire AMASS+HumanAct12 datasets? What I am trying to do is pre-process a subset of AMASS or any SMPL dataset, and pass it for inference in the H3D model. I assume that is the model that has the most general format. 

What I did was download KIT from AMASS (but this can/should be interchangeable with any AMASS subset) and pre-process the data to get new_joints, new_joints_vec, etc. Then, I run steps 1-3 to generate the .npz files ( I also change train/val/test.txt files to have only the subset samples). I am able to run all the steps fine, but finally get a vocab mismatch error when I load the model. Does this mean that I have to load the entire text dataset (and thus the whole AMASS data) to run inference?


vocab mismatch error: 

        2024-05-02 16:18:02,330 [INFO] Building dataset ... 
        Traceback (most recent call last):
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/visualizations/poses2concepts.py", line 179, in <module>
            loaded_model, train_data_loader, test_data_loader,val_data_loader = load_model_config(device=device,args=args)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/src/evaluate_m2L.py", line 56, in load_model_config
            train_data_loader, val_data_loader, test_data_loader = build_data(dataset_class=dataset_class, min_freq= min_freq,
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/loader.py", line 34, in build_data
            data_loader = dataset_class(path,path_txt=path_txt,min_freq=min_freq,joint_angles=joint_angles,
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/datasets/h3d_m2t_dataset_.py", line 21, in __init__
            data = np.load(self.original_path,allow_pickle=True)
          File "/home/dhegde/anaconda3/envs/llava/lib/python3.10/site-packages/numpy/lib/npyio.py", line 427, in load
            fid = stack.enter_context(open(os_fspath(file), "rb"))
        FileNotFoundError: [Errno 2] No such file or directory: '/home/dhegde/projects/motion-pipe/M2T-Segmentation/data/\\all_humanML3D.npz'
        (llava) dhegde@DESKTOP-26TQ9NQ:~/projects/motion-pipe/M2T-Segmentation$ python visualizations/poses2concepts.py --config ./configs_h3D/MLP.yaml --dataset_name h3D --n_map 1 --n_gifs 5 --save_results ./gifs_map_kit22_h3d
        2024-05-02 16:18:38,715 [INFO] Building dataset ... 
        2024-05-02 16:18:39,018 [INFO] Mean/Std Normalization
        9297
        2024-05-02 16:18:40,091 [INFO] CORRECT TOKENS False
        2024-05-02 16:18:40,362 [INFO] Building vocabulary with minimum frequency : 3
        2024-05-02 16:18:40,752 [INFO] The vocab size is 2958 with minimum frequency of 3 it becomes --> 1749 tokens
        2024-05-02 16:18:40,752 [INFO] Convert token to numerical values
        2024-05-02 16:18:41,040 [INFO] Using official split (No random state)
        2024-05-02 16:18:41,047 [INFO] Sort with respect to pose seq_len to reduce padding percentage --> optimize time
        2024-05-02 16:18:41,050 [INFO] DOWN SAMPLE DATA 20Hz --> 10.00 Hz
        2024-05-02 16:18:41,140 [INFO] Number  of samples VAL: 458 TRAIN: 7305, TEST: 1342 
        2024-05-02 16:18:43,219 [INFO] VOCAB SIZE  = 1749 
        2024-05-02 16:18:43,224 [INFO] Scaling 0.0625
        2024-05-02 16:18:43,228 [INFO] Applying local_recurrent attention
        Traceback (most recent call last):
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/visualizations/poses2concepts.py", line 179, in <module>
            loaded_model, train_data_loader, test_data_loader,val_data_loader = load_model_config(device=device,args=args)
          File "/home/dhegde/projects/motion-pipe/M2T-Segmentation/src/evaluate_m2L.py", line 76, in load_model_config
            loaded_model.load_state_dict(new_dict)
          File "/home/dhegde/anaconda3/envs/llava/lib/python3.10/site-packages/torch/nn/modules/module.py", line 2041, in load_state_dict
            raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
        RuntimeError: Error(s) in loading state_dict for seq2seq:
                size mismatch for dec.embedding.weight: copying a param with shape torch.Size([3605, 128]) from checkpoint, the shape in current model is torch.Size([1749, 128]).
                size mismatch for dec.fc.weight: copying a param with shape torch.Size([3605, 640]) from checkpoint, the shape in current model is torch.Size([1749, 640]).
                size mismatch for dec.fc.bias: copying a param with shape torch.Size([3605]) from checkpoint, the shape in current model is torch.Size([1749]).



Basically, I would like to be able to process any SMPL model sequence and run an inference for motion. Based on the pre-processing required, I think I should put everything in the HumanML3D format, please let me know if I am on the right track, and how I can fix the vocab size mismatch error.

**Comment by rd20karim** [comment]
Yes, it's likely the reason because the vocabulary size is derived from the textual descriptions of the entire HumanML3D dataset not only a subset of Amass which may not cover all possible used tokens (words here).

**Comment by deeptibhegde** [comment]
Is there a way to load all the textual descriptions and not the joint samples? Could you provide just the processed text of the HumanML3D dataset? i.e., the sentence_corrections.csv for HumanML3D?

**Comment by rd20karim** [comment]
Yes, it's likely possible, a quick solution you can try is to run `kit_m2t_dataset.py` using the entire corrected text and stop at this line [L49](https://github.com/rd20karim/M2T-Segmentation/blob/c7ae537fcf1d3e5741511e5a7688874c62462be1/datasets/kit_m2t_dataset.py#L49) then save the objects  [self.token_to_idx L170](https://github.com/rd20karim/M2T-Segmentation/blob/c7ae537fcf1d3e5741511e5a7688874c62462be1/datasets/vocabulary.py#L170) and L171 (`self.idx_to_token`)
Then in inference you can directly load those objects without going through `build_vocabulary` function.

I am not sure if I can give directly the textual descriptions due to the distribution policy mentioned [here](https://github.com/EricGuo5513/HumanML3D?tab=readme-ov-file#how-to-obtain-the-data)

**Comment by deeptibhegde** [comment]
If I understand correctly, I should save the objects of self.lang using the pre-processing step of KIT-ML-Aug and load them when I run inference on H3D? So during inference, I comment out [line 57 in h3d_m2t_dataset_.py](https://github.com/rd20karim/M2T-Segmentation/blob/c7ae537fcf1d3e5741511e5a7688874c62462be1/datasets/h3d_m2t_dataset_.py#L57) to and load the saved data? 

Will this give me the correct vocab size that the H3D model expects? The reason I am trying to inference using the H3D model is that when I pass the joint info to the KIT-ML-Aug model, it gives the wrong caption,  making me think maybe the format is the issue.

**Comment by rd20karim** [comment]
The KIT-ML and H3D have different number of joints each with its precise self.lang object, the KIT-ML-aug model could be used only for KIT-ML-aug data and same for H3D mainly because of difference (21 joints vs 22 joints).

Yes, you can try saving self.lang, I guess it would need some additional minor changes but it would be discovered after the first run. Certainly, this self.lang object can work only with its dataset.

**Comment by deeptibhegde** [comment]
I was able to load the model once I pre-processed the whole HumanML3D dataset. Thank you for your help.
